relay-server/src/endpoints/common.rs

```diff
@@ -286,44 +286,19 @@ fn queue_envelope(
         })
     }
 
-    // Take all NEL reports and split them up into the separate envelopes with 1 item per
-    // envelope.
-    for nel_envelope in envelope.split_all_by(|item| matches!(item.ty(), ItemType::Nel)) {
-        relay_log::trace!("queueing separate envelopes for NEL report");
-        let buffer_guard = state.buffer_guard();
-        let nel_envelope = buffer_guard
+    // Split off the envelopes by item type.
+    let envelopes = managed_envelope.take_envelope().into_processible_parts();
```

Comment:
I could do that too, I will have to add the `BufferGuard` to processor. Otherwise it wouldn't be that difficult

Reply:
Any changes we make to this have to consider the rate-limit fast path in the project cache. Since we're applying rate limits there, either we keep splitting envelopes first or we need to ensure that the rate limiting code in the project cache is aware of the dependencies between items.